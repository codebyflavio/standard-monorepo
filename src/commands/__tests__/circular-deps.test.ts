import CircularDeps from '../circular-deps'
jest.mock('../../core/allPackages')
import getAllPackages from '../../core/allPackages'
import { graphs } from '../../test/utils'

let stdout

beforeEach(() => {
  stdout = []
  jest
    .spyOn(process.stdout, 'write')
    .mockImplementation((val) => stdout.push(val))
})

describe('[command:circular-deps]', () => {
  it('should detect no circular dependencies', async () => {
    await CircularDeps.run([])
    expect(stdout).toMatchSnapshot()
  })

  it('should detect circular dependencies', async () => {
    // @ts-expect-error
    getAllPackages.mockImplementation(async () => graphs.ONE_CYCLE)

    try {
      await CircularDeps.run(['--fail'])
    } catch (err) {
      expect(err.toString()).toMatchSnapshot()
    }
  })

  it('should detect multiple circular dependencies', async () => {
    // @ts-expect-error
    getAllPackages.mockImplementation(async () => graphs.TWO_CYCLES)

    try {
      await CircularDeps.run(['--fail'])
    } catch (err) {
      expect(err.toString()).toMatchSnapshot()
    }
  })

  it('should fail silently', async () => {
    // @ts-expect-error
    getAllPackages.mockImplementation(async () => graphs.ONE_CYCLE)

    await CircularDeps.run(['--no-fail'])
    expect(stdout).toMatchSnapshot()
  })

  it('should pass if there are less circular dependencies than the maximum', async () => {
    // @ts-expect-error
    getAllPackages.mockImplementation(async () => graphs.TWO_CYCLES)

    await CircularDeps.run(['--max=5'])
    expect(stdout).toMatchSnapshot()
  })

  it('should fail if there are more circular dependencies than the maximum', async () => {
    // @ts-expect-error
    getAllPackages.mockImplementation(async () => graphs.TWO_CYCLES)

    try {
      await CircularDeps.run(['--max=1'])
    } catch (err) {
      expect(err.toString()).toMatchSnapshot()
    }
  })
})
