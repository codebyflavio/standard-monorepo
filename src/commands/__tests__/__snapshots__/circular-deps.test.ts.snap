// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[command:circular-deps] should detect circular dependencies 1`] = `
"Error: [FAIL] Found 1 circular dependencies in the project, please fix these as soon as possible.

d -> e -> f -> d"
`;

exports[`[command:circular-deps] should detect multiple circular dependencies 1`] = `
"Error: [FAIL] Found 2 circular dependencies in the project, please fix these as soon as possible.

a -> b -> c -> a

d -> e -> f -> g -> d"
`;

exports[`[command:circular-deps] should detect no circular dependencies 1`] = `
Array [
  "[SUCCESS] No circular dependencies found in the project, good job!
",
]
`;

exports[`[command:circular-deps] should fail if there are more circular dependencies than the maximum 1`] = `
"Error: [FAIL] Found 2 circular dependencies in the project, please fix these as soon as possible.

a -> b -> c -> a

d -> e -> f -> g -> d"
`;

exports[`[command:circular-deps] should fail silently 1`] = `
Array [
  "[WARNING] Found 1 circular dependencies in the project, please fix these as soon as possible.

d -> e -> f -> d
",
]
`;

exports[`[command:circular-deps] should pass if there are less circular dependencies than the maximum 1`] = `
Array [
  "[WARNING] Found 2 circular dependencies in the project, please fix these as soon as possible.

a -> b -> c -> a

d -> e -> f -> g -> d
",
]
`;
